### PR TITLE
Permanent identifier

### PR DESCRIPTION
**Pull request by maraino**
### Description

This PR adds support for the Permanent Identifier SAN (#53). With this PR, a Permanent Identifier can be added to a template data using:

```go
data.Set(x509util.SANsKey, []x509util.SubjectAlternativeName{
	{Type: x509util.PermanentIdentifierType, Value: permanentIdentifier},
})
```

It also cleans and adds more unit tests for custom SANs PR (#27), and updates `go.step.sm/crypto` CI to use Go 1.18 and Go 1.19

Note that the assigner part of the Permanent Identifier is currently not supported. Suggestions are welcome. One possible solution would be to convert the `SubjectAlternativeName.Value` to `any` and enforce string for most common types and support both string and a JSON for the Permanent Identifier. Other non-supported types like DirectoryNames can benefit from this. But if we want to do this, this will be part of a different PR.

### CODE REVIEW

File: x509util/extensions.go [hunk 1]
```diff
@@ -738,14 +754,16 @@ func MarshalSubjectAlternativeName(sans ...SubjectAlternativeName) (pkix.Extensi
 //
 // TODO(mariano,unreality): X400Address, DirectoryName, and EDIPartyName types
 // are defined in RFC5280 but are currently unimplemented
```
Comment (hslatman) [inline x509util/extensions.go:756]: The ACME DA draft RFC mentions the `hardware-module` identifier and `HardwareModule` SAN extension. I think it's the `id-on-hardwareModuleName`, as that exists on the same level of the tree as `PermanentIdentifier`. We may need to verify this is the intended one. 

`id-on-SmtpUTF8Mailbox` is an internationalized email address and defined in https://www.rfc-editor.org/rfc/rfc8398.txt. 

`id-on-dnsSRV` is a SRVName as defined in https://www.rfc-editor.org/rfc/rfc4985.html.

Not sure if the latter two make sense to add now.

**Comment by hslatman** [comment]
Your solution using `any` would result in a type switch to do something type specific when we support it, and do a transformation to a string when we don't know the type (yet)?

File: x509util/extensions.go (hunk 1 above)
Comment (maraino) [reply x509util/extensions.go:756]: AFAIK, the format of that SAN is not yet defined, but I added this https://github.com/smallstep/crypto/issues/55

File: x509util/certificate_test.go [hunk 2]
```diff
@@ -241,12 +270,13 @@ func TestNewCertificate(t *testing.T) {
...
 		{"missingTemplate", args{cr, []Option{WithTemplateFile("./testdata/missing.tpl", CreateTemplateData("commonName", []string{"foo.com"}))}}, nil, true},
 		{"badJson", args{cr, []Option{WithTemplate(`"this is not a json object"`, CreateTemplateData("commonName", []string{"foo.com"}))}}, nil, true},
+		{"failCustomSANs", args{cr, []Option{WithTemplate(DefaultLeafTemplate, badCustomSANsData)}}, nil, true},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			got, err := NewCertificate(tt.args.cr, tt.args.opts...)
 			if (err != nil) != tt.wantErr {
-				t.Errorf("NewCertificate() error = %v, wantErr %v", err, tt.wantErr)
+				t.Errorf("NewCertificate() error = %+v, wantErr %v", err, tt.wantErr)
```
[the remark this answers is not in the document]
Comment (maraino) [reply x509util/certificate_test.go]: Fixed with 33046b2

File: x509util/extensions.go [hunk 3]
```diff
@@ -796,14 +814,12 @@ func createSubjectAltNameExtension(c *Certificate, subjectIsEmpty bool) (*Extens
 	// Now marshal the rawValues into the ASN1 sequence, and create an Extension object to hold the extension
 	rawBytes, err := asn1.Marshal(rawValues)
 	if err != nil {
-		return nil, errors.Wrap(err, "error marshaling SubjectAlternativeName extension to ASN1")
+		return zero, errors.Wrap(err, "error marshaling SubjectAlternativeName extension to ASN1")
```
[the remark this answers is not in the document]
Comment (maraino) [reply x509util/extensions.go:817]: The usage of errors.Wrap in this package is all over the place; at some point we can get rid of all of them. But the stack trace capability is nice.

**Comment by maraino** [comment]
> Your solution using any would result in a type switch to do something type specific when we support it, and do a transformation to a string when we don't know the type (yet)?

Yes, most of the types would be string, and fail if they are not, but we can add custom types for more complex SANs, although at this moment there's no real need for it, so if we need to it will go to a different PR.

File: x509util/extensions.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hslatman) [reply x509util/extensions.go:817]: Al right; let's keep it like this for now 🙂

File: x509util/extensions.go (hunk 1 above)
Comment (hslatman) [reply x509util/extensions.go:756]: Agree, I'm not entirely sure about the `id-on-hardwareModuleName`at the moment. I've noted this as a question for Brandon.

**Review by hslatman — APPROVED** [review]

**Pull request merged**